model_zoo/pytorch/nanogpt/train_lightning.py

```diff
@@ -0,0 +1,387 @@
+
+    def __len__(self):
+        return len(self.data) - self.block_size
+
+    def __getitem__(self, idx):
+        x = torch.from_numpy(
+            self.data[idx : idx + self.block_size].astype(np.int64)  # noqa
+        )  # noqa
+        y = torch.from_numpy(
+            self.data[idx + 1 : idx + 1 + self.block_size].astype(  # noqa E203
+                np.int64
+            )  # noqa
+        )  # noqa
+        return x, y
+
+
+class GPTDataModule(pl.LightningDataModule):
+    def __init__(self, data_dir, batch_size, block_size):
+        self.data_dir = data_dir
+        self.batch_size = batch_size
+        self.block_size = block_size
+        self.global_step = 0
+        super().__init__()
+
+    def setup(self, stage):
+        self.train_data, self.val_data, self.vocab_size = self._get_dataset(
+            self.data_dir
+        )
+        self.train_sampler = ElasticDistributedSampler(self.train_data)
+
+    def train_dataloader(self):
+        dataset = BlockDataset(self.train_data, self.block_size)
+        return DataLoader(dataset, batch_size=self.batch_size, num_workers=4)
+
+    def val_dataloader(self):
+        dataset = BlockDataset(self.val_data, self.block_size)
+        return DataLoader(dataset, batch_size=self.batch_size, num_workers=4)
+
+    def state_dict(self):
+        steps_per_epoch = len(self.train_dataloader())
+        step_in_epoch = self.global_step % steps_per_epoch
+        epoch = int(self.global_step / steps_per_epoch)
+        self.train_sampler.set_epoch(epoch)
+        state = self.train_sampler.state_dict(step_in_epoch, self.batch_size)
+        return state
+
+    def load_state_dict(self, state_dict) -> None:
+        self.train_sampler.load_state_dict(state_dict)
+
+    def _get_dataset(self, data_dir):
+        train_data = np.memmap(
+            os.path.join(data_dir, "train.bin"), dtype=np.uint16, mode="r"
+        )
+        val_data = np.memmap(
+            os.path.join(data_dir, "val.bin"), dtype=np.uint16, mode="r"
+        )
+        # Attempt to derive vocab_size from the dataset
+        meta_path = os.path.join(data_dir, "meta.pkl")
+        meta_vocab_size = None
+        if os.path.exists(meta_path):
+            with open(meta_path, "rb") as f:
+                meta = pickle.load(f)
+        meta_vocab_size = meta["vocab_size"]
+        print(f"found vocab_size = {meta_vocab_size} (inside {meta_path})")
+        return train_data, val_data, meta_vocab_size
+
+
+def log_rank0(msg):
+    local_rank = int(os.environ.get("LOCAL_RANK", 0))
+    if local_rank == 0:
+        print(msg)
+
+
+def setup(args):
+    global local_rank, master_process
+
+    use_cuda = torch.cuda.is_available() and args.device != "cpu"
+    if use_cuda:
+        dist.init_process_group("nccl", timeout=timedelta(seconds=120))
+    else:
+        dist.init_process_group("gloo", timeout=timedelta(seconds=120))
+    rank = dist.get_rank()
+    local_rank = int(os.environ["LOCAL_RANK"])
```

Comment:
local_rank = int(os.environ.get("LOCAL_RANK", 0))

Reply:
Tks for reminding me of this line with the default value issue. I will fix it in a later commit.

Reply:
Fixed.